File: Python/Level-2(string)/reverse_words_sentence.py
def reverse_words_manual(sentence):
    # Convert string to a mutable list of characters
    chars = list(sentence)
    
    # Helper function to reverse a portion of the list in-place
    def reverse_string(start, end):
        while start < end:
            # Python allows us to swap variables in one line!
            chars[start], chars[end] = chars[end], chars[start]
            start += 1
            end -= 1

    n = len(chars)
    
    # Step 1: Reverse the entire list of characters
    reverse_string(0, n - 1)
    
    # Step 2: Reverse each individual word
    start = 0
    for i in range(n + 1):
        # If we hit a space or the end of the list, reverse the word we just passed
        if i == n or chars[i] == ' ':
            reverse_string(start, i - 1)
            start = i + 1 # Move start pointer to the next word
            
    # Convert the list of characters back into a string
    return "".join(chars)

# Test it
print(reverse_words_manual("programming in Python is fun"))

File: C/Level-2(string)/reverse_words_sentence.c
#include <stdio.h>
#include <string.h>

// Helper function to reverse characters between two pointers
void reverseString(char* start, char* end) {
    char temp;
    while (start < end) {
        // Swap the characters
        temp = *start;
        *start = *end;
        *end = temp;
        
        // Move pointers toward the center
        start++;
        end--;
    }
}

// Function to reverse the words in a sentence
void reverseWords(char* s) {
    if (s == NULL || *s == '\0') return;

    // Step 1: Reverse the entire string
    int len = strlen(s);
    reverseString(s, s + len - 1);

    // Step 2: Reverse each individual word
    char* word_start = s;
    char* temp = s;

    while (*temp) {
        temp++; // Move through the string one character at a time
        
        if (*temp == '\0') {
            // We reached the end of the string; reverse the last word
            reverseString(word_start, temp - 1);
        } else if (*temp == ' ') {
            // We hit a space; reverse the word we just passed
            reverseString(word_start, temp - 1);
            
            // Set the start of the next word to the character after the space
            word_start = temp + 1;
        }
    }
}

int main() {
    // Note: We use a char array instead of a char pointer (char *s = "...") 
    // because string literals are read-only in C and would cause a segmentation fault.
    char sentence[] = "programming in C is fun";
    
    printf("Original : %s\n", sentence);
    
    reverseWords(sentence);
    
    printf("Reversed : %s\n", sentence);
    
    return 0;
}
